Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp

```diff
@@ -0,0 +1,366 @@
+ */
+WorkspaceGroup_sptr
+LoadAndApplyMuonDetectorGrouping::addGroupedWSWithDefaultName(
+    Workspace_sptr workspace) {
+  auto &ads = AnalysisDataService::Instance();
+  std::string groupedWSName = MuonAlgorithmHelper::getRunLabel(workspace);
+
+  WorkspaceGroup_sptr groupedWS;
+  if (ads.doesExist(groupedWSName)) {
+    if ((groupedWS = boost::dynamic_pointer_cast<WorkspaceGroup>(
+             ads.retrieve(groupedWSName)))) {
+      return groupedWS;
+    }
+  }
+
+  groupedWS = boost::make_shared<WorkspaceGroup>();
+  AnalysisDataService::Instance().addOrReplace(groupedWSName, groupedWS);
+  return groupedWS;
+}
+
+// Set the parameters in options to sensible defaults for this algorithm
+AnalysisOptions LoadAndApplyMuonDetectorGrouping::setDefaultOptions() {
+  AnalysisOptions options;
+  options.summedPeriods = this->getPropertyValue("SummedPeriods");
+  options.subtractedPeriods = this->getPropertyValue("SubtractedPeriods");
+  options.timeZero = 0.0;
+  options.loadedTimeZero = this->getProperty("TimeOffset");
+  options.rebinArgs = this->getPropertyValue("RebinArgs");
+  options.plotType = Muon::PlotType::Counts;
+  return options;
+}
+
+/**
+ * Adds the group names and corresponding detector IDs to
+ * a TableWorkspace in the ADS named "MuonGroupings"
+ */
+void LoadAndApplyMuonDetectorGrouping::addGroupingInformationToADS(
+    const Mantid::API::Grouping &grouping) {
+
+  auto groupingTable = boost::dynamic_pointer_cast<ITableWorkspace>(
+      WorkspaceFactory::Instance().createTable("TableWorkspace"));
+  groupingTable->addColumn("str", "GroupName");
+  groupingTable->addColumn("vector_int", "Detectors");
+
+  size_t numGroups = grouping.groups.size();
+  for (size_t i = 0; i < numGroups; i++) {
+    TableRow newRow = groupingTable->appendRow();
+    newRow << grouping.groupNames[i];
+    std::vector<int> detectorIDs =
+        Kernel::Strings::parseRange(grouping.groups[i]);
+    std::sort(detectorIDs.begin(), detectorIDs.end());
+    newRow << detectorIDs;
+  }
+
+  AnalysisDataServiceImpl &ads = AnalysisDataService::Instance();
+  ads.addOrReplace("MuonGroupings", groupingTable);
+}
+
+/**
+ * Load the grouping and pairing information from the input file
+ * into the Grouping struct, after performing some checks.
+ */
+API::Grouping LoadAndApplyMuonDetectorGrouping::loadGroupsAndPairs() {
+  Grouping grouping;
+  std::string filename = getProperty("Filename");
+  GroupingLoader::loadGroupingFromXML(filename, grouping);
+
+  CheckValidGroupsAndPairs(grouping);
+
+  return grouping;
+}
+
+/**
+ * Check if the group/pair names are valid, and if all the groups which
+ * are paired are also included as groups.
+ */
+void LoadAndApplyMuonDetectorGrouping::CheckValidGroupsAndPairs(
+    const Grouping &grouping) {
+  for (auto &&groupName : grouping.groupNames) {
+    if (!MuonAlgorithmHelper::checkValidGroupPairName(groupName)) {
+      throw std::invalid_argument("Some group names are invalid");
```

Comment:
lets add the group name to the error message. So then the user knows what is invalid. The same with the below